src/stream/src/executor/top_n/top_n_cache.rs

```diff
@@ -159,20 +219,60 @@ impl<const WITH_TIES: bool> TopNCache<WITH_TIES> {
         }
         full
     }
+
+    /// Use this method instead of `self.high.insert` directly when possible.
+    /// It checks `is_high_cache_dirty` flag and refills the high cache if needed.
+    async fn insert_high_cache<S: StateStore>(
+        &mut self,
+        group_key: Option<impl GroupKey>,
+        managed_state: &mut ManagedTopNState<S>,
+        cache_key: CacheKey,
+        row: CompactedRow,
+    ) -> StreamExecutorResult<()> {
+        if !self.is_high_cache_full() {
+            let need_refill = self.is_high_cache_dirty
+                && match self.high.last_key_value() {
+                    Some(high_last) => cache_key > *high_last.0,
+                    None => false,
+                };
+            if need_refill {
+                managed_state
+                    .fill_high_cache(
+                        group_key,
+                        self,
+                        self.middle.last_key_value().unwrap().0.clone(),
+                        self.high_capacity,
+                    )
+                    .await?;
+                self.is_high_cache_dirty = false;
+            }
+            self.high.insert(cache_key, row);
+        } else {
+            let high_last = self.high.last_entry().unwrap();
+            if cache_key <= *high_last.key() {
+                high_last.remove_entry();
+                self.high.insert(cache_key, row);
+            }
+        }
```

Comment:
Maybe we do not need to refill the high cache here and just ignore the insert operation when `elem_to_compare_with_high.0 > *high_last.key()` so that we do not need to maintain the `is_high_cache_dirty` and the high cache will store the TopK elements forever.

Reply:
updated